### ISSUE TITLE
Queries fail when fetchOptions headers overwritten

### ISSUE THREAD

**Issue by ar1a** [FormidableLabs/urql#86]
I have overriden my fetch options like so

```
const client = new Client({
  url: 'http://localhost:4000/',
  fetchOptions: () => {
    const token = window.localStorage.getItem('token');
    console.log(token);
    if (token) {
      return {
        headers: {
          Authorization: `Bearer ${token}`
        }
      };
    }
    return {};
  }
});
```
And once token isn't null, it makes requests at plain/text instead of application/json. This is tracked down to https://github.com/FormidableLabs/urql/blob/1564d4d713059f5acf25412a7d334209d4d7b102/src/modules/client.ts#L196 in which it spreads over headers. 

I fixed this clientside by adding 
```
          'Content-Type': 'application/json'
``` 
to my fetchOptions, but this could be fixable by deep merging fetchOptions

**Comment by kitten** [comment]
It could be fixed by doing that but I’d say this is 100% expected behaviour. Since deep merging is quite unnecessary and can make it hard to override options correctly it might make more sense to just add an option for “headers”

cc @kenwheeler

**Comment by cross311** [comment]
@kitten Would it be an option to pass in the default fetch options so spread operator can be used on the modified properties?

Looking at the code looks like the defaults are hard coded, so moving up their definition up a bit (minus body) to be passed into fetchOptions if it is a function.  Then continue to do shallow spread afterwords to keep it backwards compatible.

I am game for doing a pull to take a stab at this.

**Comment by kenwheeler** [comment]
@cross311 that seems like a solid enhancement. 💯

**Comment by ar1a** [comment]
right, this should be fixed then according to #96?

**Comment by kitten** [comment]
@ar1a yep, thank you for pushing this change 👍 in retrospect it makes a lot of sense and I should’ve been less criticism. 

Also thanks @cross311 for taking care of it.

Hopefully we can land all changes soon ✨
